w3sper.js/src/network.js

```diff
@@ -0,0 +1,163 @@
+      throw new TypeError(`${url} is not a valid URL.`);
+    }
+
+    if (!["http:", "https:"].includes(url.protocol)) {
+      throw new TypeError(`${url} is not a http(s) URL.`);
+    }
+
+    const { protocol, hostname, port } = url;
+
+    Object.defineProperty(this, "url", {
+      value: new URL(`${protocol}//${hostname}` + (port ? `:${port}` : "")),
+      writable: false,
+      enumerable: true,
+    });
+  }
+
+  static connect(url, options = {}) {
+    return new Network(url).connect(options);
+  }
+
+  async connect(options = {}) {
+    const url = new URL(this.url);
+    url.protocol = protocol[url.protocol];
+    url.pathname = "/on";
+
+    const { signal } = options;
+    const socket = new WebSocket(url);
+    this.#socket = socket;
+
+    return new Promise(async (resolve, reject) => {
+      if (signal?.aborted) {
+        reject(signal.reason);
+      } else if (signal) {
+        signal.addEventListener("abort", (event) => {
+          socket.close();
+        });
+      }
+
+      await once(socket, "open");
+      const event = await once(socket, "message");
+
+      this.#sessionId = event.data;
+
+      socket.onmessage = console.log;
+
+      const url = new URL("/on/node/info", this.url);
+
+      const response = await fetch(url, {
+        method: "POST",
+      });
+
+      const info = await response.json();
+
+      const nodeInfo = Object.fromEntries(
+        Object.entries(info).map(([key, value]) => [snakeToCamel(key), value]),
+      );
+
+      this.#nodeInfo = Object.freeze(nodeInfo);
+
+      loadDriver(new URL("/static/drivers/wallet-core.wasm", this.url));
+
+      resolve(this);
+    });
+  }
+
+  async disconnect() {
+    this.#socket.close();
+    this.#nodeInfo = null;
+    this.#sessionId = undefined;
+
+    return await once(this.#socket, "close").then(
+      (event) => ((this.#socket = null), event),
+    );
+  }
+
+  get nodeInfo() {
+    return this.#nodeInfo;
+  }
+
+  get connected() {
+    return this.#socket?.readyState === WebSocket.OPEN;
+  }
+
+  get sessionId() {
+    return this.#sessionId;
+  }
+
+  // TODO: GraphQL returns a `Number` while the block height is a `BigInt`.
+  // A `Number` loses precision after 9_007_199_254_740_991, while a `BigInt`
+  // can go up to: `18_446_744_073_709_551_615`.
+  //
+  // I suspect is a GraphQL limitation. In the meantime we convert the `Number`
+  // to a `BigInt` for consistency and future proof of the API's consumers.
+  get blockHeight() {
+    return this.query("query { block(height: -1) { header { height } }}").then(
+      (body) => BigInt(body?.block?.header?.height ?? 0),
+    );
+  }
+
+  async sync(options = {}) {
+    if (!("to" in options)) {
+      options.to = await this.blockHeight;
+    }
+
+    return new Syncer(this, options);
+  }
+
+  async dispatch(resource, options = {}) {
+    const { signal } = options;
+
+    const headers = new Headers(resource.headers);
+    headers.append("rusk-version", "0.8.0");
```

Comment:
This is a server match. Should you ask for "0.8" so to connect to patched server too?

Reply:
It's a draft, it's not meant to be commented yet.